src/tasks/task_list.hpp

```diff
@@ -89,6 +164,52 @@ class TaskList {
                          [=]() mutable -> TaskStatus { return (obj->*func)(args...); });
   }
 
+  template <class F, class... Args>
+  TaskID AddIterativeTask(const TaskType &type, const std::string &label,
+                          TaskID const &dep, F &&func, Args &&... args) {
+    if (max_iterations_.count(label) == 0) {
+      max_iterations_[label] = std::numeric_limits<unsigned int>::max();
+      count_[label] = 0;
+    }
+    if (throw_with_max_iters_.count(label) == 0) {
+      throw_with_max_iters_[label] = false;
+    }
+    if (warn_with_max_iters_.count(label) == 0) {
+      warn_with_max_iters_[label] = true;
+    }
+    TaskID id(tasks_added_ + 1);
+    task_list_.push_back(Task(
+        id, dep,
+        [=, func = std::forward<F>(func)]() mutable -> TaskStatus {
+          return func(args...);
+        },
+        type, label));
+    tasks_added_++;
+    return id;
+  }
+
+  // overload to add member functions of class T to task list
+  // NOTE: we must capture the object pointer
+  template <class T, class... Args>
+  TaskID AddIterativeTask(const TaskType &type, const std::string &label,
+                          TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                          Args &&... args) {
+    return this->AddIterativeTask(
+        type, label, dep, [=]() mutable -> TaskStatus { return (obj->*func)(args...); });
+  }
+
+  void SetMaxIterations(const std::string &label, const int max) {
+    max_iterations_[label] = max;
+  }
+
+  void SetFailWithMaxIterations(const std::string &label, bool flag) {
+    throw_with_max_iters_[label] = flag;
+  }
+
+  void SetWarnWithMaxIterations(const std::string &label, bool flag) {
+    warn_with_max_iters_[label] = flag;
+  }
```

Comment:
This part of the machinery feels a little odd... in fact having iterative tasks tied to a label, rather than being a separate manager object in themselves is a little odd to me. But it does express what we want to express.

Reply:
yeah, let me think about what that looks like.

Reply:
ok @Yurlungur.  That was a little painful, but I refactored into something that perhaps looks more like you imagined, at least from the perspective of the interface.  I'm not sure whether it made the code in `task_list.hpp` more or less clear.  While I _do_ like that downstream codes won't have to provide labels and such now, my worry is that this interface obfuscates the fact that the tasks associated with some iterative collection can be interleaved with other tasks in a region that may be part of another iterative pattern or otherwise.  So, very curious to hear what you think.  Seems like both approaches have some merits.  Not at all sure which one we should stick with.  Thoughts?

Reply:
Hmm interesting. I actually like the new way it looks for applications better. I agree it somewhat obfuscates that the tasks can be interwoven, but it also makes interacting with bits in a given iterative solve more clear and intuitive I feel. So I lean towards that.

On the other hand, I really dislike the nested ness of the classes. Can we not just have them flat? Or maybe make them friends?